### ISSUE TITLE
Generating a Cpp parser for the Cpp grammar produces incorrect method call.

### ISSUE THREAD

**Issue by huperniketes** [antlr/antlr4#2116]
Xcode chokes on the generated CPP14Parser.cpp with this error:

/Users/huperniketes/Documents/Projects/Applications/Multiplatform/Cppquins/generated/CPP14Parser.cpp:16897:145: No member named 'compareTo' in 'std::__1::basic_string<char>'; did you mean 'compare'?

**Comment by ericlemes** [comment]
Bumped into the same issue here, using Antlr 4.7.0.

**Comment by niccroad** [comment]
There is a semantic predicate written in Java in that grammar. That needs to be re-written in the target language for run-times other than Java.

**Comment by ericlemes** [comment]
Yep. It is happening in all the languages. If I knew how to fix I'd be really happy so send a pull request. But I didn't understand exactly how this "templates" works in the code.

**Comment by ericvergnaud** [comment]
Looks like the cpp14 grammar contains Java productions.
Go to the google discussion group for support 


Envoyé de mon iPhone

**Comment by niccroad** [comment]
@ericlemes In your copy of the Cpp14 grammar the predicate needs to be altered from this,
`{if($val.text.compareTo("0")!=0) throw new InputMismatchException(this);}`
to
`{ if($val.getText() != "0") throw InputMismatchException(this); }`

Its under the purespecifier rule.

In terms of a pull request, there is a comment into which this fragment might be inserted for how to make this work in the Cpp target. Unfortunately the grammar predicates themselves are simply copied into the generated parsers and lexers without considering which run-time they work for. I think the active predicate should remain in Java in the Cpp14.g4 example grammar.

The grammar will 'work' without the predicate as well, but will then allow =1, =2, =3, =4, =5, =6 and =7 as alternative pure virtual function specifiers in this case.

Sorry if I made any specific mistakes in these instructions, I was not able to compile them presently.

**Comment by ericlemes** [comment]
@niccroad changing the code I managed to make it work. To be honest, I gave up what I was trying to do and started a different approach. I bumped in so many issues that kind of discouraged me to use antlr for this purpose. Perhaps for simpler grammar would work fine.

Just to summarise the kind of issue I found trying to run the C++ grammar:
- Compiling parsers generated by the antlr tool. Didn't work out of the box on C# nor C++
- Memory issues. Trying to parse a big quantity of files was using more than 4GB of memory. Since I was using the parser embedded in MSBuild, which is a 32bit process, was running out of memory.
- Tried to occasionally clear the cache (which I think should have a memory limit), bumped into performance issues. Was taking DAYS to parse the files.
- Tried a C++ parser, running on a separated process, still very slow.

I'd totally discourage to use antlr4 with big grammars until some improvements in these areas are made.

**Comment by niccroad** [comment]
@ericlemes Pragmatism is of course a good approach.

Unfortunately the Cpp14 grammar is basically copied from the C++ standard appendix which is far from an efficient form.

Another issue with this tends to be that the C pre-processor is not (and can't really be) part of the grammar. These kinds of things make general case C++ parsing hard.

**Comment by billguastalla** [comment]
The original issue still occurs for the latest version of CPP14.g4, on antlr-4.8.
